### PR TITLE
Add docs and otherwise clean up the facade crate

### PR DESCRIPTION
**Pull request by aturon**
Note that I expanded the glob re-exports because otherwise rustdoc will not actually show the contents of the modules.

### CODE REVIEW

**Comment by aturon** [comment]
Also, these docs will eventually want to have more specific pointers into the book throughout, but I'd like to do that in a separate pass once the book has more content.

File: futures/src/lib.rs [hunk 1]
```diff
@@ -1,164 +1,28 @@
...
+//! - [Futures](::Future) (sometimes called promises), which represent a single
+//! asychronous computation that may result in a final value or an error.
 //!
-//! You can find extensive tutorials and documentations at [https://tokio.rs]
-//! for both this crate (asynchronous programming in general) as well as the
-//! Tokio stack to perform async I/O with.
+//! - [Streams](::Stream), which represent a series of values or errors produced asynchronously.
 //!
-//! [https://tokio.rs]: https://tokio.rs
+//! - [Sinks](::Sink), which represent asynchronous writing of data.
```
Comment (cramertj) [inline futures/src/lib.rs]: "represent asynchronous writing of data" doesn't sound quite right to me.
"which represent asynchonous pipes into which data can be written" is too wordy, but maybe something along those lines?

File: futures/src/lib.rs [hunk 2]
```diff
@@ -1,164 +1,28 @@
...
+//! [cross-task communication](::channel).
 //!
-//! ```toml
-//! [dependencies]
-//! futures = "0.1"
-//! ```
+//! Underlying all of this is the *task system*, which is a form of lightweight
+//! threading. Large asynchronous computations are built up using futures,
+//! streams and sinks, and then spawned as independent, asynchronous tasks which
+//! run to completion, but *do not block* the thread running them.
```
Comment (cramertj) [inline futures/src/lib.rs]: [are] run to completion

File: futures/src/lib.rs [hunk 3]
```diff
@@ -1,164 +1,28 @@
...
-//! # fn main() {}
-//! ```
-//!
-//! Some more information can also be found in the [README] for now, but
-//! otherwise feel free to jump in to the docs below!
-//!
-//! [README]: https://github.com/alexcrichton/futures-rs#futures-rs
+//! **The best way to learn about this crate is through the [Asynchronous
+//! Programming in Rust](https://aturon.github.io/apr/) book**, which provides a
+//! comprehensive introduction. The docs within this crate, by contract, are
```
Comment (cramertj) [inline futures/src/lib.rs]: "by contract" seems unnecessary

File: futures/src/lib.rs [hunk 4]
```diff
@@ -192,62 +56,185 @@ macro_rules! try_ready {
     })
 }
 
-// TODO: task_local macro
+/// A macro to create a `static` of type `LocalKey`
+///
+/// This macro is intentionally similar to the `thread_local!`, and creates a
+/// `static` which has a `with` method to access the data on a task.
```
Comment (cramertj) [inline futures/src/lib.rs]: `with` was changed to [`get_mut`](https://github.com/rust-lang-nursery/futures-rs/pull/784/files#diff-48e5eeeacb6dc31a3e5b5ed5134314caR98).

File: futures/src/lib.rs (hunk 3 above)
Comment (aturon) [reply futures/src/lib.rs]: 😆 that was supposed to be "by *contrast*"

Comment (cramertj) [reply futures/src/lib.rs]: lol :)

**Comment by aturon** [comment]
Updated with new commit.

File: futures/src/lib.rs [hunk 5]
```diff
@@ -192,62 +56,185 @@ macro_rules! try_ready {
...
+
+    pub use futures_channel::{oneshot, mpsc};
 }
 
+#[cfg(feature = "std")]
 pub mod executor {
-    //! Executors
+    //! Task execution.
+    //!
+    //! All asynchronous computation occurs within an *executor*, which is
```
Comment (cramertj) [inline futures/src/lib.rs]: Lots of *italics* and "scare quotes" here. :woman_shrugging:

File: futures/src/lib.rs [hunk 6]
```diff
@@ -192,62 +56,185 @@ macro_rules! try_ready {
...
+    //! The call to [`run`](::executor::ThreadPool::run) will block the current
+    //! thread until the future defined by `my_app` completes, and will return
+    //! the result of that future.
+    //!
+    //! # Spawning additional tasks
+    //!
+    //! There are two ways to spawn a task:
+    //!
+    //! - Spawn onto a "default" execuctor by calling the top-level
+    //! [`spawn`](::executor::spawn) function.
```
Comment (cramertj) [inline futures/src/lib.rs]: Possibly worth mentioning `cx.spawn(fut)` can be used when implementing futures manually. (maybe this belongs in the docs of `spawn`, though, rather than these docs?)

File: futures/src/lib.rs [hunk 7]
```diff
@@ -192,62 +56,185 @@ macro_rules! try_ready {
...
+    //!
+    //! This module contains:
     //!
-    //! This module contains the `Future` trait and adapters for this trait.
+    //! - The [`Future` trait](::Future).
+    //!
+    //! - The [`FutureExt`](::future::FutureExt) trait, which provides adapters
+    //! for chaining and composing futures.
+    //!
+    //! - Top-level future contructors like [`lazy`](::future::lazy) which
```
Comment (cramertj) [inline futures/src/lib.rs]: s/contructors/constructors

We use the term "combinators" elsewhere

File: futures/src/lib.rs [hunk 8]
```diff
@@ -192,62 +56,185 @@ macro_rules! try_ready {
...
-    //! as a number of combinators and extensions for using them.
-    pub use futures_io::*;
-    pub use futures_util::io::*;
+    // TODO: document utility functions once the re-org is done
+
+    pub use futures_io::{
+        Error, Initializer, IoVec, ErrorKind, AsyncRead, AsyncWrite, Result
+    };
+    pub use futures_util::io::{
+        codec, io, AsyncReadExt, AsyncWriteExt
```
Comment (cramertj) [inline futures/src/lib.rs]: I think this nested `io::io` module should be gone now

**Comment by aturon** [comment]
Updated with all feedback addressed.

**Comment by cramertj** [comment]
:shipit:
